### PR TITLE
Add explicit equality checks for SupportedKinds

### PR DESCRIPTION
**Pull request by arkodg**
Fixes: https://github.com/kubernetes-sigs/gateway-api/issues/1703

Signed-off-by: Arko Dasgupta <arko@tetrate.io>

### CODE REVIEW

**Comment by shaneutt** [comment]
Hey :wave: I had assigned myself this and was working on an implementation, however I'm very happy to differ to you.

Two extra things I was doing:

1. adding unit tests. This whole problem would have been avoided if we had unit tests (which, to be clear, is on us as maintainers)
2. add an error condition for duplicate SupportedKinds: technically it's never "correct" to have duplicates in this list, so this is a place we can helpfully let implementations know if they're doing something weird.

Looks like you already started writing some unit tests, perhaps to test the original issue. I don't mean to put extra work on you, but if you wouldn't mind adding unit tests for the relevant functionality I think that can help us from having problems here again. If you're not up for this, please do say so I'm happy to take this on.

In any case, very much appreciate the help, thank you! Let me know how I can be helpful to you :vulcan_salute:

**Comment by skriss** [comment]
Can confirm this fixes the issue for Contour, thanks @arkodg!

**Comment by arkodg** [comment]
> Hey 👋 I was working on an implementation of this as well, however I'm very happy to differ to you.
> 
> Two extra things I was doing:
> 
> 1. adding unit tests. This whole problem would have been avoided if we had unit tests (which, to be clear, is on us as maintainers)
> 2. add an error condition for duplicate SupportedKinds: technically it's never "correct" to have duplicates in this list, so this is a place we can helpfully let implementations know if they're doing something weird.
> 
> Looks like you already started writing some unit tests, perhaps to test the original issue. I don't mean to put extra work on you, but if you wouldn't mind adding unit tests for the relevant functionality I think that can help us from having problems here again. If you're not up for this, please do say so I'm happy to take this on.
> 
> In any case, very much appreciate the help, thank you! Let me know how I can be helpful to you 🖖

hi @shaneutt , ive added tests to this PR, haven't handled the duplicate case in this PR, please feel free to take care of it in a future PR

File: conformance/utils/kubernetes/helpers.go
Comment (shaneutt) [inline conformance/utils/kubernetes/helpers.go:647]: It's a bit of a nit, but in multiple places here we're de-referencing `Group` and the attribute is technically optional and _could_ be `nil`.

We colloquially expect any `RouteGroupKinds` that come through here will have been initialized via the API first so generally we wouldn't expect this to be an issue because of CRD defaults that would be applied. That said, some situation where someone later tries to use these helper functions in a more derived way would leave them with a segmentation fault that might be little terse.

One thing you could do would be just initialize them with the default value before using them if needed:

```diff
diff --git a/conformance/utils/kubernetes/helpers.go b/conformance/utils/kubernetes/helpers.go
index a0926f9d..33a0c7a8 100644
--- a/conformance/utils/kubernetes/helpers.go
+++ b/conformance/utils/kubernetes/helpers.go
@@ -638,6 +638,14 @@ func listenersMatch(t *testing.T, expected, actual []v1beta1.ListenerStatus) boo
                for _, eKind := range eListener.SupportedKinds {
                        found := false
                        for _, aKind := range aListener.SupportedKinds {
+                               if eKind.Group == nil {
+                                       eKind.Group = (*v1beta1.Group)(&v1beta1.GroupVersion.Group)
+                               }
+
+                               if aKind.Group == nil {
+                                       aKind.Group = (*v1beta1.Group)(&v1beta1.GroupVersion.Group)
+                               }
+
                                if *eKind.Group == *aKind.Group && eKind.Kind == aKind.Kind {
                                        found = true
                                        break
```

File: conformance/utils/kubernetes/helpers_test.go
Comment (shaneutt) [inline conformance/utils/kubernetes/helpers_test.go:12]:
```suggestion
func Test_listenersMatch(t *testing.T) {
```

File: conformance/utils/kubernetes/helpers_test.go
Comment (robscott) [inline conformance/utils/kubernetes/helpers_test.go:29]: Do these ever actually vary? Can we either omit them entirely or use something like a `commonConditions` var?

File: conformance/utils/kubernetes/helpers_test.go
Comment (robscott) [inline conformance/utils/kubernetes/helpers_test.go:34]: Same question for name - is this actually needed?

**Review by robscott — COMMENTED** [review]
Thanks @arkodg!

File: conformance/utils/kubernetes/helpers_test.go
Comment (shaneutt) [inline conformance/utils/kubernetes/helpers_test.go:157]: One thing I noticed is that because we only test for `bool` return values, and we don't have a mechanism here to verify the logging output our checking of results is a bit limited. That said however, this isn't something I think we need to resolve right here and now because it would grow the scope of this PR considerably to refactor this now, and this is blocking the `v0.6.1` release. As such, as per @mikemorris' comments throughout, I'll make a follow-up issue to do some refactoring in here in a later iteration.

**Review by shaneutt — COMMENTED** [review]
No hard blockers, though I do have a few comments to be resolved before we merge.

Thank you so much for jumping in and adding this fix @arkodg we appreciate you! :vulcan_salute: 

cc @robscott / @youngnick to take a look as well.

File: conformance/utils/kubernetes/helpers_test.go
Comment (shaneutt) [inline conformance/utils/kubernetes/helpers_test.go:13]: Somehow I had a comment here that got dropped from my previous review.

While looking at coverage for this, there are a few cases we're not covering:

```golang
        if len(expected) != len(actual) {
                t.Logf("Expected %d Gateway status listeners, got %d", len(expected), len(actual))
                return false
        }
```

```golang
                if aListener.Name != eListener.Name {
                        t.Logf("Name doesn't match")
                        return false
                }
```

```golang
                if aListener.AttachedRoutes != eListener.AttachedRoutes {
                        t.Logf("Expected AttachedRoutes to be %v, got %v", eListener.AttachedRoutes, aListener.AttachedRoutes)
                        return false
                }
```

```golang
                if !conditionsMatch(t, eListener.Conditions, aListener.Conditions) {
                        t.Logf("Expected Conditions to be %v, got %v", eListener.Conditions, aListener.Conditions)
                        return false
                }
```

e.g.:

`go test -coverprofile cover.out ./conformance/utils/kubernetes/...`
`go tool cover -html=cover.out -o cover.html`
`firefox cover.html`

File: conformance/utils/kubernetes/helpers_test.go
Comment (shaneutt) [reply conformance/utils/kubernetes/helpers_test.go:157]: https://github.com/kubernetes-sigs/gateway-api/issues/1708 was created to resolve this comment :+1:

File: conformance/utils/kubernetes/helpers_test.go
Comment (arkodg) [reply conformance/utils/kubernetes/helpers_test.go:13]: yah, these should be tracked with a GH issue and fixed in a follow up PR :)

File: conformance/utils/kubernetes/helpers_test.go
Comment (arkodg) [reply conformance/utils/kubernetes/helpers_test.go:12]: can make this change, but hope it doesnt mean all test cases associated with `listenersMatch` need to be handled in this PR :)

File: conformance/utils/kubernetes/helpers.go
Comment (arkodg) [reply conformance/utils/kubernetes/helpers.go:647]: since `Group` is optional, and if you believe GAPI code might skip populating `Group`, then it makes more sense for the `Group` check to be skipped imho

File: conformance/utils/kubernetes/helpers_test.go
Comment (arkodg) [reply conformance/utils/kubernetes/helpers_test.go:34]: sure, will rm

File: conformance/utils/kubernetes/helpers_test.go
Comment (arkodg) [reply conformance/utils/kubernetes/helpers_test.go:29]: sure will rm

**Comment by arkodg** [comment]
tested this PR in https://github.com/envoyproxy/gateway/pull/990

File: conformance/utils/kubernetes/helpers.go
Comment (shaneutt) [reply conformance/utils/kubernetes/helpers.go:647]: I don't think _skipping_ the group check would be right, because if you did that you could have two kinds of the same name, with two different group names that would be considered matching right?

File: conformance/utils/kubernetes/helpers_test.go
Comment (shaneutt) [reply conformance/utils/kubernetes/helpers_test.go:13]: If this is pushing it for your available time, then I can agree to making it a follow-up as what we have here would appear to be a significant improvement over the previously broken logic. Thank you for your time so far :bow:

Comment (shaneutt) [reply conformance/utils/kubernetes/helpers_test.go:13]: Added to https://github.com/kubernetes-sigs/gateway-api/issues/1708 :+1:

File: conformance/utils/kubernetes/helpers.go
Comment (arkodg) [reply conformance/utils/kubernetes/helpers.go:647]: also previously the check was an exact Match and all implementations abided by it, will make this change if thats your preference

Comment (shaneutt) [reply conformance/utils/kubernetes/helpers.go:647]: Unless I've missed something, this just means we'll provide the already configured default behavior in edge cases rather than a segfault, so yeah if it's not a problem I would appreciate it thank you.

Comment (arkodg) [reply conformance/utils/kubernetes/helpers.go:647]: sg, will add

**Review by shaneutt — APPROVED** [review]
Looking good.

cc @robscott or @youngnick for the final LGTM please.

**Comment by robscott** [comment]
Thanks @arkodg!

/lgtm
/ok-to-test

File: conformance/utils/kubernetes/helpers_test.go
Comment (robscott) [inline conformance/utils/kubernetes/helpers_test.go:1]: This file needs a boilerplate header, for example https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/doc.go#L1-L15C3 (but with 2023 as the year)

Comment (arkodg) [reply conformance/utils/kubernetes/helpers_test.go:1]: ack

**Comment by shaneutt** [comment]
@arkodg looks like imports need cleanup:

```
conformance/utils/kubernetes/helpers_test.go:19: File is not `goimports`-ed with -local sigs.k8s.io/gateway-api (goimports)
```

File: conformance/utils/kubernetes/helpers_test.go
Comment (arkodg) [reply conformance/utils/kubernetes/helpers_test.go:14]: the tool still thinks its 2022, will raise a GH issue for it

**Comment by arkodg** [comment]
> @arkodg looks like imports need cleanup:
> 
> ```
> conformance/utils/kubernetes/helpers_test.go:19: File is not `goimports`-ed with -local sigs.k8s.io/gateway-api (goimports)
> ```

thanks fixed it, there was also an issue around the date mentioned in LICENSE - `hack/verify-boilerplate.sh` didn't like it when I set the year to `2023`, set it to `2022` for now

**Comment by robscott** [comment]
Thanks @arkodg!

/lgtm

**Pull request merged**
